### DISCUSSION TITLE
port to STM32 and W5500

### DISCUSSION THREAD

**Discussion by JAndrassy** [ssilverman/QNEthernet#62] [Show and tell]
Hi Shawn

You know I do a research on Arduino networking libraries. But I don't have all the specific hardware so I try to run the LwIP based libraries on boards I have. I made the Ethernet library for Portenta C33 Eth peripheral run on an Uno R4 with an Ethernet shield. I adopted the hw interfacing functions to a driver for W5500 from esp8266 Arduino (LwIP too). Then I used the same trick with the STM32Ethernet library, which was written for a STM32 MCU with an Eth peripheral. I used Black Pill, a board with STM32 F411 MCU without the Eth peripheral. 

And then I ported your QNEthernet to the STM32 core and W5500 and I tested it on the same Black Pill with a W550 module.  Everything I wanted to test worked.

The port to STM32:
* I copied elapsedMillis.h from Teensy core into the library
* I removed  use of EventResponder (and patched a yield function implementation in the sketch)
* I just commented out Entrophy in RandomDevice (I don't want to test secure connection)
* in sys_arch I used millis() in sys_now implementation (why systick_millis_count?)
* in lwipopts.h I removed `#include <imxrt.h>` and `#include <sys/time.h>`. they seem unused.
* in ConnectionManager I removed `#include <core_pins.h>`. seems unused too
* avr/progmem in STM32 core doesn't have DMAMEM and FLASHMEM so I just used `-DDMAMEM="" -DFLASHMEM=""`
* I hade to -DTEENSYDUINO=160, because the IPAddress operators from ip_tools were in conflict with IPAddress 

For the W5500 port I wrote lwip_w5500.cpp. I attach it. All the s_netif and netif lines are duplicated from lwip_t41.c. Maybe do a compare. What was possible I left in place.
For the w5500 I had to change lwipopts.h. Enable the checksums, because the W5500 doesn't do them and set ETH_PAD_SIZE to 0. 

[lwip_w5500.cpp](https://github.com/ssilverman/QNEthernet/files/13781416/lwip_w5500.cpp.txt)

The yield function I used instead of EventResponder
```
void yield(void) {
  static bool busy = false;
  if (!busy && Ethernet) {
    busy = true;
    Ethernet.loop();
    busy = false;
  }
}
```

**Comment by ssilverman** [comment]
Thanks for testing the waters here. I've made a bunch of changes that should address the issues you saw during porting:
1. No more `elapsedMillis` dependencies.
2. `EventResponder` is now gated on the `ARDUINO_TEENSY41` and `ARDUINO_TEENSY40` macros, and provides a place to put your own yield function or handler. See just below the `attachLoopToYield()` function in _QNEthernet.cpp_.
3. `RandomDevice` now works on non-Teensy boards by using `std::rand()` (may be improved later).
4. _sys_arch.cpp_ now defines `sys_now()` to use `millis()`.
5. Fixed _lwipopts.h_ dependency on `<imxrt.h>` and `<sys/time.h>` for non-Teensy boards. Also improved ETH_PAD_SIZE and checksums for non-Teensys.
6. Removed `<core_pins.h>` include from `ConnectionManager`.
7. For those platforms without DMAMEM and FLASHMEM, fixed their use.
8. The _ip_tools_ files should work better now.

In other words, all the issues should be solved and the library should be able to be used out-of-the-box with STM32. I tested with `blackpill_f411ce` in PlatformIO. I haven't yet pushed my changes, so watch for those soon.

I also wrote my own W5500 driver that I'll share in the near future.

Other changes of note:
1. Revamped driver interface.
2. New _qnethernet_opts.h_ file for changing _QNEthernet_ options without having to set project-wide macros.

Note: Random numbers aren't just used for crypto or security. They're for these things (non-exhaustive list):
1. Initializing `std::rand()` with a seed,
2. Selecting local ports when opening TCP and UDP sockets,
3. mDNS response delays,
4. Random delays for other protocols to help the device cooperate on a network, etc.

**Comment by JAndrassy** [comment]
Hello. what a nice surprise. I am glad I inspired you. 
so now all ARM Teensy are supported with W5500? And STM32 with W5500?
the STM32 Arduino community has a forum https://www.stm32duino.com. Let them know when it is ready.

I am working on integrating a driver for ESP32 as WiFi adapter. Espressif has [a firmware](https://github.com/espressif/esp-hosted) for ESP32 which makes it an MAC level WiFi adapter. Arduino made a driver for the host side for their new Portenta C33 board. It has a form of an Arduino library and I extracted it from their Renesas Core into a stand-alone [repository](https://github.com/JAndrassy/ESPHost) and I made it compatible with any ARM Arduino (originally it used Reneasas SDK for SPI, delay and pin manipulation).

Now I finished [a first version](https://github.com/JAndrassy/arduino-pico/tree/lwip_esphost/libraries/lwIP_ESPHost) of integrating the ESPHost library int the RP2040 core, which uses LwIP from RP2040 SDK and has WiFi only for Pico W on-board WiFi. The maintainer added support for wired Ethernet with W5500 and enc28j60, but there was no WiFi option for other RP2040 boards .

**Comment by ssilverman** [comment]
I just pushed the latest `master`. Let me know what you think. I haven't yet committed the W5500 driver, but here's my current version. To use it, un-comment `QNETHERNET_DRIVER_W5500` in _qnethernet_opts.h_ (the new version attached here). You can unzip the file in the root of the _QNEthernet_ library. (Don't forget to update/pull first.)

* v1: [w5500_driver.zip](https://github.com/ssilverman/QNEthernet/files/14060781/w5500_driver.zip)
* v2: [w5500_driver.zip](https://github.com/ssilverman/QNEthernet/files/14060876/w5500_driver.zip)
* v3: [w5500_driver.zip](https://github.com/ssilverman/QNEthernet/files/14071331/w5500_driver.zip)

(Use the latest; the earlier versions are historical but only slightly different.)

I'm sure there's room for improvement in the design and in the W5500 driver, but I think it's a good start for:
1. Compiling for other platforms, and
2. Adding new drivers.

> so now all ARM Teensy are supported with W5500? And STM32 with W5500?

Yes, correct. (Well, it _should_, but I haven't yet tested with a Teensy 4.0.)

**Comment by ssilverman** [comment]
The upcoming 0.27.0 release will contain the new driver. Note that it’s now in `master`. I tested with Teensy 3.1 and 4.1.

**Comment by JAndrassy** [comment]
sorry, I still work on WiFi drivers in Pico Core when I have time for Arduino.
do you plan to add the enc28j60 driver too? people still use with these powerful MCUs the EthernetENC library which was optimized for ATmega.
and maybe the WiFi drivers I make for Pico Core could be ported to your library later.

**Comment by ssilverman** [comment]
Oh, I was just adding the detail that I added my W5500 driver to the library. This is just informational; there’s no need or expectation to do anything on your end. Note also that I added a new driver interface API.

If I had some hardware, the time, and the will, I could add a driver for the ENC28J60, I suppose. Who would use it?

Side question: If you have the time and the willingness, would you be willing to test the latest on an STM32 with the new W5500 driver? Simply uncomment the line that defines `QNETHERNET_DRIVER_W5500` in _qnethernet_opts.h_.

**Reply by JAndrassy** [reply]
I maintain the EthernetENC library and it collects GitHub stars

**Comment by JAndrassy** [comment]
in `EthernetClass::start()` I had to move `driver_set_chip_select_pin(chipSelectPin_);` before `if (!driver_has_hardware()) {`.
then it works

**Reply by ssilverman** [reply]
Great catch. Fixed and pushed. Thanks!

**Comment by JAndrassy** [comment]
just for fun I attempted to compile the library with the Arduino SAMD core. but they are on C++ 11 version of arm gcc, so no luck. I was lucky then choosing the  STM32 Core

**Reply by ssilverman** [reply]
Which SAMD board are you using? I got it compiling with C++11, but the demo main program appears too large for an Arduino Zero.

**Reply by JAndrassy** [reply]
the goal was to make it build for one of Adafruit's M4 SAMD51 boards. the Adafruit SAMD core is a fork of the Arduino Core but it supports SAMD51

boards URL  `https://adafruit.github.io/arduino-board-index/package_adafruit_index.json`

**Reply by ssilverman** [reply]
It seems to build properly for `adafruit_feather_m4`, but I don't have hardware to test it with.

**Reply by JAndrassy** [reply]
and I was curious if it could fit into the M0/Zero. so it doesn't

**Reply by ssilverman** [reply]
I'll push some changes shortly for you to try...

**Reply by JAndrassy** [reply]
I can buy Adafruit ItsyBitsy M4 for 20 Euro here, so I could test it.

**Reply by JAndrassy** [reply]
and a new Arduino is coming.this year. a Nano with a Cortex-C33 Silicon Labs MCU. it doesn't have LwIP bundled with Core.

**Reply by ssilverman** [reply]
I just tested a build and, theoretically, it seems to fit the ItsyBitsy M4. See the new `c++11` branch.

Note that I built with these options (obviously, the raw frame support is optional):
1. `QNETHERNET_DRIVER_W5500`
2. `QNETHERNET_ENABLE_RAW_FRAME_SUPPORT=1`

**Reply by JAndrassy** [reply]
it works on the Adafruit M4

**Reply by ssilverman** [reply]
That’s great news! Did you run some of the examples or something?

**Reply by JAndrassy** [reply]
yesterday I didn't have much time. I only run the LegacyEthernetTest. today evening I will do more testing.

**Reply by JAndrassy** [reply]
so it doesn't work. DHCP works and then nothing. connect returns true, but doesn't really connect. no packets in Wireshark.
but it is not a problem of the C++11 branch or the SAMD51. it doesn't work on STM32 with the master branch too.

**Reply by JAndrassy** [reply]
my bad. it works. on the M4 too. I forgot to add:
```
void yield(void) {
  static bool busy = false;
  if (!busy && Ethernet) {
    busy = true;
    Ethernet.loop();
    busy = false;
  }
}
```
I thought it was no longer necessary

**Reply by ssilverman** [reply]
Cool!

**Reply by ssilverman** [reply]
That’s odd you had to add that. Do you have the latest version?
See: https://github.com/ssilverman/QNEthernet/blob/6813ca565275feb3b1db5a5d15e104189b4cd7f3/src/QNEthernet.cpp#L57

**Reply by ssilverman** [reply]
Are you saying you needed to implement `yield()` _and_ `HAS_EVENT_RESPONDER` is set? I made an assumption that if that macro is set then you don’t need `yield()`. Maybe that isn’t true?

**Reply by JAndrassy** [reply]
sorry. no. yeld() is there. I will investigate why it wasn't applied

**Reply by JAndrassy** [reply]
`yield` in QNEthernet.cpp is in the namespace so it is not linked to the global declaration of `yield()`

**Reply by ssilverman** [reply]
Fixed. Thanks for noticing that.

**Comment by ssilverman** [comment]
Note: I’ve been making efforts to keep the `c++11` branch up to date.
